### DISCUSSION TITLE
The future: JSPI and unsafe access handles

### DISCUSSION THREAD

**Discussion by rhashimoto** [rhashimoto/wa-sqlite#115] [General]
# The teaser
How would you like a persistent browser database that...
* ...doesn't have the time and space penalties of Asyncify?
* ...doesn't have COOP/COEP restrictions?
* ...works from multiple contexts, e.g. from multiple tabs?
* ...supports simultaneous read transactions?
* ...supports write-ahead logging (WAL)?
* ...moves data faster than IndexedDB?
* ...might have similar write transaction overhead to IndexedDB?
* ...allows database copying in/out and renaming using the OPFS API?

# The proposals
There are two web platform proposals of particular interest in the pipeline.

## WebAssembly JavaScript Promise Integration (JSPI)
[JSPI](https://github.com/WebAssembly/js-promise-integration/blob/main/proposals/js-promise-integration/Overview.md) is basically a replacement for the [original Asyncify](https://kripken.github.io/blog/wasm/2019/07/16/asyncify.html) that patched WASM to unwind and rewind the stack to accommodate asynchronous calls from WASM to JavaScript. The original Asyncify was clever and amazing, but had substantial size and performance penalties, especially with SQLite. JSPI uses stack switching to provide the same programming bridge, but without the penalties. This makes it a lot easier to connect SQLite with asynchronous JavaScript APIs (like OPFS, IndexedDB, Web Locks, etc.) without making painful tradeoffs.

JSPI is working now in Chrome 104+ behind a flag (#enable-experimental-webassembly-stack-switching).

## Locking modes for file access handles
New [locking modes for FileSystemSyncAccessHandle](https://github.com/whatwg/fs/blob/main/proposals/MultipleReadersWriters.md) allow multiple access handles to be opened on the same file. Previously an OPFS file could have only one open access handle at a time. The key addition is new locking mode "readwrite-unsafe", which works a lot more like the POSIX file handles that C programs use. This makes it a lot easier and more performant to support multiple connections to a SQLite database on OPFS.

The locking modes are working now in Chrome 120+ (now in the Canary channel) behind a flag (#file-system-access-locking-scheme).

# The prototype
Together, these proposals provide a straightforward path to a fully featured SQLite virtual file system on OPFS. I built SQLite replacing Asyncify with JSPI, and implemented an OPFS VFS in [this wa-sqlite branch](https://github.com/rhashimoto/wa-sqlite/tree/jspi+readwrite-unsafe). For the purposes of this prototype, my goal was to get something working as quickly as possible so I hardwired the VFS rather than rewrite all the wa-sqlite glue that allows implementing VFS extensions in JavaScript. The code is in [libopfs.c](https://github.com/rhashimoto/wa-sqlite/blob/jspi%2Breadwrite-unsafe/src/libopfs.c) and [libopfs.js](https://github.com/rhashimoto/wa-sqlite/blob/jspi%2Breadwrite-unsafe/src/libopfs.js), and here's the [Makefile diff](https://github.com/rhashimoto/wa-sqlite/compare/master...jspi+readwrite-unsafe#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52).

Currently the main demo (and probably nothing else) works on Chrome Canary with the required flags enabled. The hardest part was really just figuring out how to get the first JSPI call to work - everything after that was straightforward because the need for clever workarounds has magically vanished. I only implemented exclusive locking for now but there's no barrier to supporting the full locking model. I'm pretty sure that the shared memory methods needed for WAL can be added but I have not attempted it.

I'm using EMSDK 3.1.45 and Chrome 120.0.6053.0 (Official Build) canary (arm64). Note that because these proposals are not final, the combination of my code, Emscripten, and Chrome may be incompatible at different versions (e.g. I had to inject a [small workaround](https://github.com/rhashimoto/wa-sqlite/blob/fbc17c651dfe82409b4b6fb88556625bd7acc7a2/demo/demo-worker.js#L26-L31) because the latest EMSDK isn't pre-built for my platform).

# The takeaway
Proposals are subject to change or abandonment, the approval process can be long, and universal browser support can take even longer. Nevertheless, there seems to be good momentum here. I feel hopeful that this means we'll finally get the WebSQL replacement everyone wanted.

One little detail: a key reason why building this prototype was so easy is that the wa-sqlite API was designed to accommodate Asyncify from the start. In particular, any API call that _could_ have an asynchronous JavaScript call in its call graph is itself defined to be asynchronous. This fortuitously meant that swapping JSPI for Asyncify had almost no impact on both public and internal APIs. That might not be the case with other libraries - for example, the official SQLite WASM project treats all its WASM exports as synchronous - and that may cause async contagion headaches in this particular future for those libraries and applications that use them.

**Comment by maccman** [comment]
Incredible!

**Comment by rhashimoto** [comment]
> How would you like a persistent browser database that...
> 
> * ...doesn't have the time and space penalties of Asyncify?
> * ...doesn't have COOP/COEP restrictions?
> * ...works from multiple contexts, e.g. from multiple tabs?
> * ...supports simultaneous read transactions?
> * ...supports write-ahead logging (WAL)?
> * ...moves data faster than IndexedDB?
> * ...might have similar write transaction overhead to IndexedDB?
> * ...allows database copying in/out and renaming using the OPFS API?

I built a follow-on proof-of-concept SQLite VFS that largely achieves all these goals. A demo is online [here](https://rhashimoto.github.io/floor-demo/demo/), with instructions [here](https://github.com/rhashimoto/wa-sqlite/discussions/116#discussioncomment-7376715).

**Comment by rhashimoto** [comment]
There's good news and bad news with JSPI. The good news is that the Chrome implementation doesn't crash so much on me, and [Firefox has begun development](https://bugzilla.mozilla.org/show_bug.cgi?id=1850627). The bad news is now that it's stable enough to run benchmarks, Chrome seems to have a significant performance issue, at least on my platform (M2 Mac). Asynchronous calls from WebAssembly to JavaScript are faster with JSPI than Asyncify, but in the other direction, from JS to WASM, JSPI is horrendously slower. Hundreds of times slower. Here are comparison benchmarks with the same VFS:

![Screenshot 2024-01-14 at 5 25 49 PM](https://github.com/rhashimoto/wa-sqlite/assets/156154/97b96dc0-1efb-4d3b-bfe7-22849bc1ff35)
I filed a [Chromium bug](https://bugs.chromium.org/p/chromium/issues/detail?id=1518374&q=&can=4) (I botched the bug title, hope they will fix that). I also filed an [Emscripten bug](https://github.com/emscripten-core/emscripten/issues/21081) even though I thought it was more likely to be a Chromium problem - it indeed appears that Emscripten is not to blame but there has been some engagement there with knowledgable people.

The news on FileSystemSyncAccessHandle readwrite-unsafe locks is that they reach Chrome stable in a few weeks.

**Reply by rhashimoto** [reply]
My [Chromium bug](https://bugs.chromium.org/p/chromium/issues/detail?id=1518374) got a developer response! It sounds like JSPI calls into WebAssembly are unavoidably more expensive than JSPI calls from WebAssembly, but there may be ways to bring that cost down. Please +1 the bug if you have an interest in this - that is one of the inputs Chromium uses to guide development priorities (*don't* post a "Me, too" comment as that just annoys them).

**Reply by rhashimoto** [reply]
Firefox has a new [tracking bug for JSPI](https://bugzilla.mozilla.org/show_bug.cgi?id=1897981), and it can be enabled with about:config flag "javascript.options.wasm_js_promise_integration" in Nightly. As of 2024-06-15 it already seems to work pretty well (better than Chrome did this early).

Chromium hasn't visibly attempted to address their [JSPI performance bug](https://issues.chromium.org/issues/41491351) but they did recently elevate it to Priority P1.

**Reply by rhashimoto** [reply]
> Chromium hasn't visibly attempted to address their [JSPI performance bug](https://issues.chromium.org/issues/41491351) but they did recently elevate it to Priority P1.

Looks fixed on today's Chrome Canary. The Chrome team anticipates [JSPI will be released this calendar year](https://chromestatus.com/feature/5674874568704000).

Judging by both the Chrome and Firefox implementations so far, JSPI doesn't look like a huge performance winner over Asyncify for SQLite calls made from Javascript. On benchmarks it looks faster on some and slower on others:

![Screenshot 2024-07-03 at 5 13 21 PM](https://github.com/rhashimoto/wa-sqlite/assets/156154/c7f77d4f-bc61-4eb7-8c4c-b378ab4c980b)

Bear in mind that these benchmarks don't model an actual application very well.

JSPI WebAssembly modules are smaller than Asyncify, about half the size uncompressed (closer to two-thirds the size compressed):

```bash
$ ls -l dist/*.wasm
-rwx------   1 roy            staff     1223567 2024-06-16 10:44 dist/wa-sqlite-async.wasm
-rwx------   1 roy            staff      602830 2024-06-16 10:44 dist/wa-sqlite-jspi.wasm
-rwx------   1 roy            staff      595826 2024-06-16 10:44 dist/wa-sqlite.wasm
```

**Reply by rhashimoto** [reply]
It looks like [Apple has started implementing JSPI in WebKit](https://github.com/WebKit/WebKit/pull/52673)! So it's [stable in Chrome](https://chromestatus.com/feature/5674874568704000), [behind a flag in Firefox](https://github.com/rhashimoto/wa-sqlite/discussions/115#discussioncomment-9775278), and apparently [coming to Safari](https://github.com/WebKit/standards-positions/issues/422#issuecomment-3321209076).

**Comment by rhashimoto** [comment]
[JSPI is a focus area for Interop 2026](https://web.dev/blog/interop-2026#jspi_for_wasm). This means that attaining a standards-compliant implementation is a priority for all the major browsers this year.
